src/java.base/share/classes/sun/security/tools/keytool/Main.java

```diff
@@ -1922,22 +1978,56 @@ private void doGenKeyPair(String alias, String dname, String keyAlgName,
         }
 
         PrivateKey privKey = keypair.getPrivateKey();
+        CertificateExtensions ext;
 
-        CertificateExtensions ext = createV3Extensions(
-                null,
-                null,
-                v3ext,
-                keypair.getPublicKeyAnyway(),
-                null);
+        if (signerFlag) {
+            Certificate signerCert = keyStore.getCertificate(signerAlias);
+
+            X509CertImpl certImpl;
+            if (signerCert instanceof X509CertImpl) {
+                    certImpl = (X509CertImpl) signerCert;
+            } else {
+                certImpl = new X509CertImpl(signerCert.getEncoded());
+            }
+
+            KeyIdentifier signerSubjectKeyId = certImpl.getSubjectKeyId();
+            if (signerSubjectKeyId == null) {
+                signerSubjectKeyId = new KeyIdentifier(signerCert.getPublicKey());
+            }
+
+            ext = createV3Extensions(
+                    null,
+                    null,
+                    v3ext,
+                    keypair.getPublicKeyAnyway(),
+                    signerSubjectKeyId);
+        } else {
+            ext = createV3Extensions(
+                    null,
+                    null,
+                    v3ext,
+                    keypair.getPublicKeyAnyway(),
+                    null);
+        }
 
         X509Certificate[] chain = new X509Certificate[1];
```

Comment:
Since the chain might contain one, I'd suggest we just declare a `newCert` here. When signer flag is not on, we can simply get the chain with `new Certificate[] {newCert}`.

Reply:
Not sure the reason why a change is needed for the existing logic.

Reply:
With a signer, it makes no sense to create a single-cert array at the beginning. I am suggesting:
```
X509Certificate newCert  = keypair.getSelfCertificate(...);
Certificate[] finalChain;
if (signerFlag) {
    finalChain = new ...
    finalChain[0] = newCert;
} else {
   finalChain = new Certificate[] { newCert };
}
keyStore.setEntry(..., finalChain);
```

Reply:
Done.